compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/IR/CUDNNTypes.cpp

```diff
@@ -8,34 +8,42 @@
 
 #include "openxla/compiler/nvgpu/Dialect/CUDNN/IR/CUDNNTypes.h"
 
+#include <cstdint>
+#include <variant>
+
+#include "llvm/ADT/TypeSwitch.h"
+#include "mlir/IR/AffineMap.h"
 #include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypeInterfaces.h"
 #include "mlir/IR/DialectImplementation.h"
+#include "mlir/Support/LogicalResult.h"
 #include "openxla/compiler/nvgpu/Dialect/CUDNN/IR/CUDNNDialect.h"
-#include "llvm/ADT/TypeSwitch.h"
 
 using namespace mlir;
 using namespace openxla::compiler::nvgpu::cudnn;
 
-static LogicalResult parseDimensionList(AsmParser &parser,
-                                        ::llvm::SmallVector<int64_t> &dims,
-                                        Type &type) {
-  llvm::SmallVector<int64_t> res;
-  Type resT;
-  if (!succeeded(parser.parseDimensionList(res, /*allowDynamic=*/true,
-                                           /*withTrailingX=*/true)) ||
-      !succeeded(parser.parseType(resT)))
+static ParseResult parseDimensionList(AsmParser &parser,
+                                      SmallVector<int64_t> &dims, Type &type) {
+  if (failed(parser.parseDimensionList(dims, /*allowDynamic=*/true,
+                                       /*withTrailingX=*/true)) ||
+      failed(parser.parseType(type)))
     return failure();
-  dims = std::move(res);
-  type = std::move(resT);
   return success();
 }
 
 static void printDimensionList(AsmPrinter &printer, ArrayRef<int64_t> dims,
                                Type type) {
-  llvm::interleave(dims, printer.getStream(), "x");
-  printer << 'x' << type;
+  auto print = [&](int64_t dim) {
+    ShapedType::isDynamic(dim) ? printer << "?" : printer << dim;
+  };
+  llvm::interleave(dims, printer.getStream(), print, "x");
```

Comment:
Total nit, but wouldn't

`for (auto dim : dims) printer << dim << "x";`

be easier to read?